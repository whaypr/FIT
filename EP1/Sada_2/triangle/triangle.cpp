#include <iostream>
#include <vector>
#include <string>
#include <cmath>

using namespace std;


class Matrix {
public:
	Matrix() {}
	Matrix(int w, int h): data( vector<vector<int>>(w, vector<int>(h)) ) {}

	int & operator () (int x, int y) { return data[x][y]; }

private:
	vector<vector<int>> data;
};


class Triangle {
public:
	Triangle(int h): height(h), height_mantinels(h+1) {}

	void load() {
		triangle = Matrix(2 * height_mantinels - 1, height_mantinels);
		possible_triangles = Matrix(2 * height_mantinels - 1, height_mantinels);


		// read actual triangle
		for (int y = 1; y <= height; ++y) {
			for (int x = y; x <= 2 * height - y; ++x) {
				char unit;
				cin >> unit;
				triangle(x, y) = unit == '#' ? 0 : 1;
			}
		}
	}

	int getSubTriangleMaxArea() {
		int max_triangle = 0;

		// triangles poiting down (as orig tr)
		for (int y = 1; y <= height; ++y) {
			for (int x = y; x <= height*2 - y; x += 2) { 
		 		if ( !triangle(x, y) ) 
		  			possible_triangles(x, y) = 0; 
		 		else if ( !triangle(x, y-1) ) {
		  			possible_triangles(x, y) = 1; 
		  			if (max_triangle == 0) max_triangle++;
		 		} else {
		 			int val = min( possible_triangles(x-1, y-1), possible_triangles(x+1, y-1) ) + 1;
		 			possible_triangles(x, y) = val; 
		 			max_triangle = val > max_triangle ? val : max_triangle;
		 		}
			}
		}

		// triangles poiting up
		for (int y = height - 1; y >= 1; --y) {
			for (int x = y+1; x <= height*2 - y; x += 2) { 
		 		if ( !triangle(x, y) ) 
		  			possible_triangles(x, y) = 0; 
		 		else if ( !triangle(x, y+1) ) {
		  			possible_triangles(x, y) = 1;
		  			if (max_triangle == 0) max_triangle++;
		 		} else {
		 			int val = min( possible_triangles(x-1, y+1), possible_triangles(x+1, y+1) ) + 1;
		 			possible_triangles(x, y) = val; 
		 			max_triangle = val > max_triangle ? val : max_triangle;
		 		}
			}
		}

		// calculate actual area
		int area = 0;
		for (int i = 2 * max_triangle - 1; i >= 1; i -= 2)
			area += i;

		return area;
	}

private:
	int height;
	int height_mantinels;
	Matrix triangle; // whole triangle
	Matrix possible_triangles; // in each unit triangle remember the biggest triangle it can end in it
};


int main() {
	// iterate through triangles
	for (int triangle = 1; ; ++triangle) {
		// get triangle height
		int height;
		cin >> height;
		if (height == 0) break; // end of input

		// find the biggest triangle
		Triangle tr(height);
		tr.load();
		int area = tr.getSubTriangleMaxArea();

		// print result
		cout << "Triangle #" << triangle << endl;
		cout << "The largest triangle area is " << area << "." << endl;
		cout << endl;
	}

	return 0;
}
