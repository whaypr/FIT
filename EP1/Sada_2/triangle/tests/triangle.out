Triangle #1
The largest triangle area is 1.

Triangle #2
The largest triangle area is 0.

Triangle #3
The largest triangle area is 1.

Triangle #4
The largest triangle area is 4.

Triangle #5
The largest triangle area is 1.

Triangle #6
The largest triangle area is 4.

Triangle #7
The largest triangle area is 1.

Triangle #8
The largest triangle area is 4.

Triangle #9
The largest triangle area is 9.

Triangle #10
The largest triangle area is 0.

Triangle #11
The largest triangle area is 25.

Triangle #12
The largest triangle area is 9.

Triangle #13
The largest triangle area is 16.

Triangle #14
The largest triangle area is 25.

Triangle #15
The largest triangle area is 25.

Triangle #16
The largest triangle area is 25.

Triangle #17
The largest triangle area is 36.

Triangle #18
The largest triangle area is 25.

Triangle #19
The largest triangle area is 36.

Triangle #20
The largest triangle area is 36.

Triangle #21
The largest triangle area is 36.

Triangle #22
The largest triangle area is 529.

Triangle #23
The largest triangle area is 169.

Triangle #24
The largest triangle area is 64.

Triangle #25
The largest triangle area is 36.

Triangle #26
The largest triangle area is 10000.

Triangle #27
The largest triangle area is 0.

Triangle #28
The largest triangle area is 7921.

Triangle #29
The largest triangle area is 7569.

Triangle #30
The largest triangle area is 4761.

Triangle #31
The largest triangle area is 8281.

Triangle #32
The largest triangle area is 8100.

Triangle #33
The largest triangle area is 5929.

Triangle #34
The largest triangle area is 5776.

Triangle #35
The largest triangle area is 9801.

Triangle #36
The largest triangle area is 7921.

Triangle #37
The largest triangle area is 9801.

Triangle #38
The largest triangle area is 9801.

Triangle #39
The largest triangle area is 9801.

Triangle #40
The largest triangle area is 8281.

Triangle #41
The largest triangle area is 9216.

Triangle #42
The largest triangle area is 9409.

Triangle #43
The largest triangle area is 9801.

Triangle #44
The largest triangle area is 7921.

Triangle #45
The largest triangle area is 9604.

Triangle #46
The largest triangle area is 9604.

Triangle #47
The largest triangle area is 9801.

Triangle #48
The largest triangle area is 9604.

Triangle #49
The largest triangle area is 8836.

Triangle #50
The largest triangle area is 9801.

Triangle #51
The largest triangle area is 9801.

Triangle #52
The largest triangle area is 5476.

Triangle #53
The largest triangle area is 5476.

Triangle #54
The largest triangle area is 9604.

Triangle #55
The largest triangle area is 9604.

Triangle #56
The largest triangle area is 7225.

Triangle #57
The largest triangle area is 5329.

Triangle #58
The largest triangle area is 6084.

Triangle #59
The largest triangle area is 7056.

Triangle #60
The largest triangle area is 6400.

Triangle #61
The largest triangle area is 9409.

Triangle #62
The largest triangle area is 9409.

Triangle #63
The largest triangle area is 9801.

Triangle #64
The largest triangle area is 5476.

Triangle #65
The largest triangle area is 5476.

Triangle #66
The largest triangle area is 9604.

Triangle #67
The largest triangle area is 9025.

Triangle #68
The largest triangle area is 5776.

Triangle #69
The largest triangle area is 5476.

Triangle #70
The largest triangle area is 5625.

Triangle #71
The largest triangle area is 5776.

Triangle #72
The largest triangle area is 7056.

Triangle #73
The largest triangle area is 6084.

Triangle #74
The largest triangle area is 6241.

Triangle #75
The largest triangle area is 6400.

Triangle #76
The largest triangle area is 6561.

Triangle #77
The largest triangle area is 6724.

Triangle #78
The largest triangle area is 6561.

Triangle #79
The largest triangle area is 6561.

Triangle #80
The largest triangle area is 7225.

Triangle #81
The largest triangle area is 7396.

Triangle #82
The largest triangle area is 5929.

Triangle #83
The largest triangle area is 9801.

Triangle #84
The largest triangle area is 7396.

Triangle #85
The largest triangle area is 7396.

Triangle #86
The largest triangle area is 7225.

Triangle #87
The largest triangle area is 7225.

Triangle #88
The largest triangle area is 9409.

Triangle #89
The largest triangle area is 5476.

Triangle #90
The largest triangle area is 5625.

Triangle #91
The largest triangle area is 9409.

Triangle #92
The largest triangle area is 9604.

Triangle #93
The largest triangle area is 5184.

Triangle #94
The largest triangle area is 6724.

Triangle #95
The largest triangle area is 9409.

Triangle #96
The largest triangle area is 9409.

Triangle #97
The largest triangle area is 9216.

Triangle #98
The largest triangle area is 9216.

Triangle #99
The largest triangle area is 9025.

Triangle #100
The largest triangle area is 9025.

Triangle #101
The largest triangle area is 8836.

Triangle #102
The largest triangle area is 8836.

Triangle #103
The largest triangle area is 8649.

Triangle #104
The largest triangle area is 8649.

Triangle #105
The largest triangle area is 8464.

Triangle #106
The largest triangle area is 8464.

Triangle #107
The largest triangle area is 8281.

Triangle #108
The largest triangle area is 9801.

Triangle #109
The largest triangle area is 9604.

Triangle #110
The largest triangle area is 9604.

Triangle #111
The largest triangle area is 9409.

Triangle #112
The largest triangle area is 9409.

Triangle #113
The largest triangle area is 9216.

Triangle #114
The largest triangle area is 9216.

Triangle #115
The largest triangle area is 9025.

Triangle #116
The largest triangle area is 9025.

Triangle #117
The largest triangle area is 8836.

Triangle #118
The largest triangle area is 8836.

Triangle #119
The largest triangle area is 8649.

Triangle #120
The largest triangle area is 6400.

Triangle #121
The largest triangle area is 7569.

Triangle #122
The largest triangle area is 9801.

Triangle #123
The largest triangle area is 9604.

Triangle #124
The largest triangle area is 9604.

Triangle #125
The largest triangle area is 9409.

Triangle #126
The largest triangle area is 9409.

Triangle #127
The largest triangle area is 9216.

Triangle #128
The largest triangle area is 9216.

Triangle #129
The largest triangle area is 9025.

Triangle #130
The largest triangle area is 9025.

Triangle #131
The largest triangle area is 8836.

Triangle #132
The largest triangle area is 8836.

Triangle #133
The largest triangle area is 8649.

Triangle #134
The largest triangle area is 8649.

Triangle #135
The largest triangle area is 8464.

Triangle #136
The largest triangle area is 8464.

Triangle #137
The largest triangle area is 8281.

Triangle #138
The largest triangle area is 7396.

Triangle #139
The largest triangle area is 7569.

Triangle #140
The largest triangle area is 7744.

Triangle #141
The largest triangle area is 7921.

Triangle #142
The largest triangle area is 7921.

Triangle #143
The largest triangle area is 10000.

Triangle #144
The largest triangle area is 8464.

Triangle #145
The largest triangle area is 7569.

Triangle #146
The largest triangle area is 9801.

Triangle #147
The largest triangle area is 9801.

Triangle #148
The largest triangle area is 9604.

Triangle #149
The largest triangle area is 9604.

Triangle #150
The largest triangle area is 7569.

Triangle #151
The largest triangle area is 9216.

Triangle #152
The largest triangle area is 9025.

Triangle #153
The largest triangle area is 9409.

Triangle #154
The largest triangle area is 9604.

Triangle #155
The largest triangle area is 8836.

Triangle #156
The largest triangle area is 8649.

Triangle #157
The largest triangle area is 8649.

Triangle #158
The largest triangle area is 9801.

Triangle #159
The largest triangle area is 5625.

Triangle #160
The largest triangle area is 5776.

Triangle #161
The largest triangle area is 6561.

Triangle #162
The largest triangle area is 6400.

Triangle #163
The largest triangle area is 5476.

Triangle #164
The largest triangle area is 7396.

Triangle #165
The largest triangle area is 9604.

Triangle #166
The largest triangle area is 5776.

Triangle #167
The largest triangle area is 8100.

Triangle #168
The largest triangle area is 7921.

Triangle #169
The largest triangle area is 7921.

Triangle #170
The largest triangle area is 7744.

Triangle #171
The largest triangle area is 7744.

Triangle #172
The largest triangle area is 7569.

Triangle #173
The largest triangle area is 7569.

Triangle #174
The largest triangle area is 9801.

Triangle #175
The largest triangle area is 10000.

Triangle #176
The largest triangle area is 6724.

Triangle #177
The largest triangle area is 6889.

Triangle #178
The largest triangle area is 7056.

Triangle #179
The largest triangle area is 7225.

Triangle #180
The largest triangle area is 7396.

Triangle #181
The largest triangle area is 7569.

Triangle #182
The largest triangle area is 7744.

Triangle #183
The largest triangle area is 7921.

Triangle #184
The largest triangle area is 5184.

Triangle #185
The largest triangle area is 7056.

Triangle #186
The largest triangle area is 7056.

Triangle #187
The largest triangle area is 9216.

Triangle #188
The largest triangle area is 4761.

Triangle #189
The largest triangle area is 8281.

Triangle #190
The largest triangle area is 8100.

Triangle #191
The largest triangle area is 5929.

Triangle #192
The largest triangle area is 5776.

Triangle #193
The largest triangle area is 9801.

Triangle #194
The largest triangle area is 7921.

Triangle #195
The largest triangle area is 9801.

Triangle #196
The largest triangle area is 9801.

Triangle #197
The largest triangle area is 9801.

Triangle #198
The largest triangle area is 8281.

Triangle #199
The largest triangle area is 9216.

Triangle #200
The largest triangle area is 4761.

Triangle #201
The largest triangle area is 8281.

Triangle #202
The largest triangle area is 8100.

Triangle #203
The largest triangle area is 5929.

Triangle #204
The largest triangle area is 5776.

Triangle #205
The largest triangle area is 9801.

Triangle #206
The largest triangle area is 7921.

Triangle #207
The largest triangle area is 9801.

Triangle #208
The largest triangle area is 9801.

Triangle #209
The largest triangle area is 9801.

Triangle #210
The largest triangle area is 8281.

Triangle #211
The largest triangle area is 9216.

